ludwig/benchmarking/reporting.py

```diff
@@ -33,3 +38,152 @@ def create_metrics_report(experiment_name: str) -> Tuple[Dict[str, Any], str]:
     merged_file_path = os.path.join(os.getcwd(), experiment_name, "metrics_report", "{}.json".format("full_report"))
     save_json(merged_file_path, full_report)
     return full_report, merged_file_path
+
+
+def initialize_stats_dict(main_function_events: List[profiler_util.FunctionEvent]) -> Dict[str, Any]:
+    """Initialize dictionary which stores resource usage information per tagged code block.
+
+    :param main_function_events: list of main function events.
+    """
+    info = dict()
+    for event_name in [evt.name for evt in main_function_events]:
+        info[event_name] = {"code_block_tag": event_name}
+        info[event_name]["runs"] = []
+    return info
+
+
+def get_memory_details(kineto_event: _KinetoEvent) -> Tuple[str, int]:
+    """Get device name and number of bytes (de)allocated during an event.
+
+    :param kineto_event: a Kineto event instance.
+    """
+    if kineto_event.device_type() in [DeviceType.CPU, DeviceType.MKLDNN, DeviceType.IDEEP]:
+        return "cpu", kineto_event.nbytes()
+    elif kineto_event.device_type() in [DeviceType.CUDA, DeviceType.HIP]:
+        return f"cuda_{kineto_event.device_index()}", kineto_event.nbytes()
+    else:
+        raise ValueError(f"Device {kineto_event.device_type()} is not valid.")
+
+
+def get_device_memory_usage(
+    kineto_event: _KinetoEvent, mem_records_acc: profiler_util.MemRecordsAcc, run_usage_info: Dict[str, Any]
+) -> Dict[str, Any]:
+    """Get CPU and CUDA memory usage for an event.
+
+    :param kineto_event: a Kineto event instance.
+    :param mem_records_acc: memory events wrapped in a data structure for fast lookup.
+    :param run_usage_info: usage info for one execution of the tagged code block.
+    """
+    records_in_interval = mem_records_acc.in_interval(
+        kineto_event.start_us(), kineto_event.start_us() + kineto_event.duration_us()
+    )
+    memory_so_far = defaultdict(int)
+    count_so_far = defaultdict(int)
+    average_so_far = defaultdict(float)
+    max_so_far = defaultdict(int)
+
+    for mem_record in records_in_interval:
+        device, nbytes = get_memory_details(mem_record[0])
+        memory_so_far[device] += nbytes
+        max_so_far[device] = max(max_so_far[device], memory_so_far[device])
+        average_so_far[device] = (memory_so_far[device] + (average_so_far[device] * count_so_far[device])) / (
+            count_so_far[device] + 1
+        )
+        count_so_far[device] += 1
+    for device in count_so_far:
+        run_usage_info[f"average_{device}_memory_usage"] = average_so_far[device]
+        run_usage_info[f"max_{device}_memory_usage"] = max_so_far[device]
+    return run_usage_info
```

Comment:
Why are we computing the metrics statefully? Is it expensive to iterate through `records_in_interval`?

```python
devices = set([get_memory_details(mem_record[0])[0] for mem_record in records_in_interval])

memory_usage = defaultdict(list)
for mem_record in records_in_interval:
    device, nbytes = get_memory_details(mem_record[0])
    memory_usage[device].append(nbytes)

max_memory_usage = defaultdict(int)
average_memory_usage = defaultdict(int)
for device in devices:
    max_memory_usage[device] = max(memory_usage[device])
    average_memory_usage[device] = mean(memory_usage[device])
...
```

Reply:
Yes, it is expensive. I made the change from `.append`, then `max` and `mean` after timing this function.

As an example, when we profile a `ludwig.experiment` on the `ames_housing` dataset (1,460 examples) with `train_step: 5` which will take 0.87 seconds, `len(records_in_interval) = 54372`. Doing `.append`, `max` and `mean` for longer jobs will be pretty expensive.